### DISCUSSION TITLE
119.29.29.29查询小众的域名时间长，223.5.5.5 IP更新结果慢，怎么能综合它们的优势？

### DISCUSSION THREAD

**Discussion by yyysuo** [pymumu/smartdns#1094] [Ideas]
比如在国内组，把119.29.29.29设置为可信dns，当223.5.5.5返回IP结果快时，设置一个3s的ttl，等119.29.29.29返回结果时，采纳119.29.29.29的查询结果。

**Comment by PikuZheng** [comment]
阿里云和腾讯云的dns表现不同原因在于，阿里云做了一个巨大的缓存，当域名缓存ttl到期后依旧向客户端提供ttl为0的缓存结果（与smartdns的“Serve expired”行为类似）。腾讯云在缓存过期后，会继续向上游发起查询直到获取到最新可信结果，然后再答复下游。这就造成了阿里云应答快，但在（尤其是ddns之类的，包括自家域名解析服务）域名过期后仍答复错误的结果；腾讯应答慢，但结果准确。

考虑到阿里的https支持问题，我建议弃用阿里云。对于smartdns开启缓存和预读，可以获得和阿里云相当的性能。

**Reply by yyysuo** [reply]
我能理解，一直也是用腾讯云的，不过一些小众的网站的响应时间达到1000ms以上后就有点影响体验了，所以才想着是不是能解决一下。
PS：您最新编译的[1.2022.v38.0.7](https://github.com/PikuZheng/smartdns/releases/tag/1.2022.v38.0.7)在x86_64的openwrt上跑不起来，luci显示未运行。

**Reply by PikuZheng** [reply]
我openwrt好像还没自动升那个版本。。。是带-check-edns 的版本还是不带的？

>小众的网站的响应时间达到1000ms以上

都1秒了早就超时了 啥网站能举个例子吗

**Reply by yyysuo** [reply]
是不带-check-edns的版本
wget https://github.com/PikuZheng/smartdns/releases/download/1.2022.v38.0.7/luci-app-smartdns.1.2022.v38.0.7.all-luci-all.ipk
wget https://github.com/PikuZheng/smartdns/releases/download/1.2022.v38.0.7/smartdns.1.2022.v38.0.7.x86_64-optware-all.ipk

网站示例，可能访问一次有缓存之后就不会有1000ms这么长时间了。
www.aihoom.com
www.efvcms.com

**Reply by PikuZheng** [reply]
openwrt不能用optware的版本

**Reply by yyysuo** [reply]
不好意思，我的错误，眼花了。

**Reply by PikuZheng** [reply]
用的doh.pub北京节点，www.aihoom.com 首次查询 897ms，之后250左右；阿里云辽宁节点，首次711ms，之后140左右。
感觉区别不大。

**Reply by yyysuo** [reply]
接近1s的表现其实是有所感知了，223.5.5.5查询也就是8ms，感觉值得搞一搞.

**Reply by PikuZheng** [reply]
>223.5.5.5查询也就是8ms

首次不会这么快的。。只不过阿里缓存比较久

**Reply by yyysuo** [reply]
![全板dns流程+ MOSDNS](https://user-images.githubusercontent.com/10081071/193071711-ee6a5cff-4a9e-4eb7-8aa7-c0e39c10f314.jpg)


使用了另类的办法实现了我的需求，目前我在用流程图中的方案2，做了如下改动：
smartdns国内上游设置了119.29.29.29、223.5.5.5、Adguard Home 2（上游119.29.29.29，开乐观缓存，可信dns）一共3个dns，主Adguard Home 1设置ttl为3s，smartdns设置response-mode fastest-response。

这样当一个小众域名第1次访问时，如果223.5.5.5最快且与119.29.29.29查询结果一致，那么体验是最优的，第2次及后续访问有正确IP的缓存可使用；如果第1次访问时223.5.5.5返回了过期的IP且响应最快，但是119.29.29.29和Adguard Home 2返回的正确结果慢，第1次访问失败；第2次及后续访问时，因为Adguard Home 2中已经有了乐观缓存，响应的时间在1ms以下，那么正确的IP就会被采用。
